tools/nl/embeddings/finetune.py

```diff
@@ -0,0 +1,258 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Fine Tune the base model."""
+
+from datetime import datetime
+import glob
+import os
+import time
+from typing import Any, List
+
+from absl import app
+from absl import flags
+from google.cloud import storage
+import gspread
+import pandas as pd
+from sentence_transformers import InputExample
+from sentence_transformers import losses
+from sentence_transformers import SentenceTransformer
+from torch.utils.data import DataLoader
+import utils
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_string('model_name_v2', 'all-MiniLM-L6-v2', 'Model name')
+flags.DEFINE_string('bucket_name_v2', 'datcom-nl-models', 'Storage bucket')
+flags.DEFINE_string('embeddings_size', '',
+                    'Valid values are: "small", "medium"')
+
+flags.DEFINE_string(
+    'autogen_input_basedir', 'data/autogen_input',
+    'Base path for CSVs with autogenerated SVs with name and description. '
+    'The actual path is `{--autogen_input_base}/{--embeddings_size}/*.csv`.')
+
+flags.DEFINE_string('alternatives_filepattern', 'data/alternatives/*.csv',
+                    'File pattern (relative) for CSVs with alternatives')
+
+flags.DEFINE_string('sentence_pairs_filepath',
+                    'data/finetuning/sentence_pairs.csv',
+                    'File path for sentence pairs CSV')
+
+BATCH_SIZE = 256
+NUM_WARMUP_STEPS = 10
+NUM_EPOCHS = 10
+
+VERY_HIGH_MATCH_SCORE = 0.95
+HIGH_MATCH_SCORE = 0.90
+MEDIUM_HIGH_MATCH_SCORE = 0.85
+
+
+def _upload_to_gcs(ctx: utils.Context,
+                   gcs_path: str,
+                   local_filepath: str,
+                   empty_folder: bool = False) -> None:
+  print(f'Uploading {local_filepath}')
+  blob = ctx.bucket.blob(gcs_path)
+
+  if empty_folder:
+    blob.upload_from_string(
+        '', content_type='application/x-www-form-urlencoded;charset=UTF-8')
+  else:
+    blob.upload_from_filename(local_filepath)
+  print(f'Path in GCS: {gcs_path}')
+
+
+def _make_gcs_model_folder(base_model_name: str) -> str:
+  now = datetime.now()
+
+  month_str = utils.two_digits(now.month)
+  day_str = utils.two_digits(now.day)
+  hour_str = utils.two_digits(now.hour)
+  minute_str = utils.two_digits(now.minute)
+  second_str = utils.two_digits(now.second)
+
+  return f"model_{base_model_name}_{now.year}_{month_str}_{day_str}_{hour_str}_{minute_str}_{second_str}"
```

Comment:
how about s/model_/finetuned_/ ?

Reply:
Added a TODO comment. This will get updated in the followup PR where the finetuned model name in GCS will be different